File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/java/src/main/java/com/example/graphalgorithms/BipartiteMatching.java
package com.example.graphalgorithms;

import java.util.*;

public class BipartiteMatching {
    public static Map<String, String> maximumMatching(Map<String, List<String>> graph) {
        Map<String, String> matchRight = new HashMap<>();
        for (String left : graph.keySet()) {
            tryMatch(left, graph, new HashSet<>(), matchRight);
        }
        Map<String, String> result = new HashMap<>();
        for (Map.Entry<String, String> entry : matchRight.entrySet()) {
            result.put(entry.getValue(), entry.getKey());
        }
        return result;
    }

    private static boolean tryMatch(
        String left,
        Map<String, List<String>> graph,
        Set<String> seen,
        Map<String, String> matchRight
    ) {
        for (String right : graph.getOrDefault(left, List.of())) {
            if (!seen.add(right)) {
                continue;
            }
            if (!matchRight.containsKey(right) || tryMatch(matchRight.get(right), graph, seen, matchRight)) {
                matchRight.put(right, left);
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        System.out.println("Bipartite matching");
        System.out.println(maximumMatching(Map.of("u1", List.of("v1", "v2"), "u2", List.of("v1"), "u3", List.of("v2", "v3"))));
    }
}

File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/python/graph_algorithms/bipartite_matching.py
from __future__ import annotations


def bipartite_matching(graph: dict[str, list[str]]) -> dict[str, str]:
    match_right: dict[str, str] = {}

    def can_match(left: str, seen: set[str]) -> bool:
        for right in graph.get(left, []):
            if right in seen:
                continue
            seen.add(right)
            if right not in match_right or can_match(match_right[right], seen):
                match_right[right] = left
                return True
        return False

    for left in graph:
        can_match(left, set())
    return {left: right for right, left in match_right.items()}


if __name__ == "__main__":
    sample = {"u1": ["v1", "v2"], "u2": ["v1"], "u3": ["v2", "v3"]}
    print("Bipartite matching")
    print(bipartite_matching(sample))

